Sources/flintc/Compiler.swift

```diff
@@ -16,26 +16,99 @@ import Optimizer
 import IRGen
 
 /// Runs the different stages of the compiler.
-struct Compiler {
-  var inputFiles: [URL]
-  var stdlibFiles: [URL]
-  var outputDirectory: URL
-  var dumpAST: Bool
-  var emitBytecode: Bool
-  var diagnostics: DiagnosticPool
-
-  var sourceContext: SourceContext {
-    return SourceContext(sourceFiles: inputFiles)
+public struct Compiler {
+  private var diagnoseOnly: Bool
+  private var inputFiles: [URL]?
+  private var stdlibFiles: [URL]?
+  private var outputDirectory: URL?
+  private var dumpAST: Bool?
+  private var emitBytecode: Bool?
+  private var diagnostics: DiagnosticPool?
+
+  private var sourceContext: SourceContext? {
+    if let inputFiles = inputFiles {
+      return SourceContext(sourceFiles: inputFiles)
+    }
+
+    return nil
+  }
+
+  /**
+   * Constructs a Compiler that supports only diagnosis of input files
+   */
+  public init(inputFiles: [URL]) {
+    self.diagnoseOnly = true
+
+    self.inputFiles = inputFiles
+  }
+
+  /**
+   * Constructs a fully fledged Compiler
+   */
+  public init(inputFiles: [URL], stdlibFiles: [URL], outputDirectory: URL,
+       dumpAST: Bool, emitBytecode: Bool, diagnostics: DiagnosticPool) {
+    self.diagnoseOnly = false
+
+    self.inputFiles = inputFiles
+    self.stdlibFiles = stdlibFiles
+    self.outputDirectory = outputDirectory
+    self.dumpAST = dumpAST
+    self.emitBytecode = emitBytecode
+    self.diagnostics = diagnostics
   }
 
   func tokenizeFiles() throws -> [Token] {
+    guard let inputFiles = inputFiles else {
+      return []
+    }
+
     let stdlibTokens = try StandardLibrary.default.files.flatMap { try Lexer(sourceFile: $0, isFromStdlib: true).lex() }
     let userTokens = try inputFiles.flatMap { try Lexer(sourceFile: $0).lex() }
 
     return stdlibTokens + userTokens
   }
 
-  func compile() throws -> CompilationOutcome {
+  public func diagnose() throws -> [Diagnostic] {
```

Comment:
Rather than copying the entire compile function can we not have a compiler initialised with that flag just behave differently when we call compile? The flag then disables file output or whatever